File: Data Structures & Algorithms/binary-search/submission-2.java
class Solution {
    public int search(int[] nums, int target) {
        int l = 0;
        int r = nums.length-1;
        
        while (l <= r){
            int mid = (int)((l+r)/2);
            System.out.println("Current index at " + mid);
            if (nums[mid] < target){
                l = mid+1;
            }else if (nums[mid] > target){
                r = mid-1;
            }else if (nums[mid] == target){
                return mid;
            }
        }
        return -1;
    }
}

File: Data Structures & Algorithms/combination-sum-iv/submission-2.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        # return the number of possible combinations that add up to target
        # you can reuse numbers
        # sort the array first so we reach base cases faster

        # break out if the current sum is greater than target or if there's no more to inspect
        # at each step we need to explore all unchosen
        # this means we need to maintain a visited bool array

        # at each step we must take this value
        # we have the choice for extension
        # we can either continue with the current value
        # or jump to a different value
        # meaning we just need to loop over everything
        # explore all options that are unvisited
        nums = sorted(nums)
        

        # think of it as if you're trying to fill in the gap

        memo = {}
        # store what you can retry also
        def search(remaining):
            if remaining in memo:
                return memo[remaining]
            if remaining == 0:
                return 1
            
            # re-explore curr
            combinations = 0
            for num in nums:
                if num > remaining:
                    break
                combinations += search(remaining-num)
            
            
            memo[remaining] = combinations
            return combinations
        
        return search(target)
            
                